### PR TITLE
[generate] Make all methods and logits processors agnostic to lm_head output size

### PR DESCRIPTION
**Pull request by Cyrilvallez**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48846) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48846)

# What does this PR do?

As per the title. This is mostly needed for Apertus in https://github.com/huggingface/transformers/pull/47662#issuecomment-5667280115, where the size of the embedding is different than the size of the lm_head output.

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -406,10 +406,21 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        # In some very rare cases, the embedding vocab may be larger than the lm_head output - try to recover in those cases by padding
+        # the logits (the tokens with id > output_vocab cannot be generated by the model, so there's no need to apply a penalty to them)
+        except RuntimeError:
+            vocab_size = scores.shape[-1]
+            scores = torch.nn.functional.pad(scores, (0, input_ids.max() + 1 - vocab_size))
+            score = torch.gather(scores, 1, input_ids)
+            score = torch.where(score < 0, score * self.penalty, score / self.penalty)
+            scores_processed = scores.scatter(1, input_ids, score)
+            # Drop whatever we padded, since thoses indices are outside the output vocab they cannot be generated so no need to apply penalty
+            scores_processed = scores_processed[..., :vocab_size]
```
Comment (vasqu) [inline src/transformers/generation/logits_process.py]: I'd rather we always pad no? Or calculate the pad size and make an if, I hate catching exceptions moreso if its a runtime exception lol

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -3288,6 +3288,7 @@ def _get_top_k_continuations(
         else:
             topk_log_probs, topk_indices = torch.topk(accumulated_log_probs, k=beams_to_keep)
 
+        vocab_size = accumulated_log_probs.shape[-1] // num_beams
```
Comment (vasqu) [inline src/transformers/generation/utils.py:3291]: this is nice

**Review by vasqu — APPROVED** [review]
Approval on the premise we remove the try catch :D (and a test ideally)

File: src/transformers/generation/logits_process.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/generation/logits_process.py]: Yeah, I was going back and forth between the two approches tbh. Not a fan of an `if`, as it's data-dependent and will necessarily create a sync with cpu, but we can always pad and clamp instead (should not be much of an overhead tbh)

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -406,10 +406,21 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         if input_ids.dim() == 1:
             input_ids = input_ids.unsqueeze(1)
 
+        # In some very rare cases, the embedding vocab may be larger than the lm_head output - in those cases, clamp the
+        # ids > output vocab_size to a fix index that we drop afterwards (the tokens with id > output_vocab cannot be
+        # generated by the model, so there's no need to apply a penalty to them)
+        vocab_size = scores.shape[-1]
+        input_ids = input_ids.clamp(max=vocab_size)
+        # Pad by 1 to absorb all the indices larger than output vocab that were clamped to the vocab
+        scores = torch.nn.functional.pad(scores, (0, 1))
```
Comment (Cyrilvallez) [inline src/transformers/generation/logits_process.py:415]: Carefully benchmarked this, and it's at least 2 order of magnitudes faster than a single decode forward for llama3 8B, so should not bring worst perfs at all

**Comment by Cyrilvallez** [comment]
@vasqu I ran the following quick benchmark on gpu and cpu, and it's 2 order of magnitudes faster than a decode forward for llama3 8B on cuda, 3 order of magnitudes faster on cpu, so I believe it's all fine with always padding even when we don'0t need it
Note that I took relatively large seq_length/vocab_size, and the difference will be even larger with smaller values

```python
import torch
from transformers import AutoModelForCausalLM
import time

N_repeat = 50
vocab_size = 100000
batch_size = 5
sequence_length = 10000
device = "cpu"

ids = torch.randint(0, vocab_size, (batch_size, sequence_length), dtype=int, device=device)
scores = torch.rand((batch_size, vocab_size), device=device)

times = []
for i in range(N_repeat):
    if torch.device(device).type == "cuda":
        torch.cuda.synchronize()
    t0 = time.time()
    foo = ids.clamp(max=vocab_size)
    # Pad by 1 to absorb all the indices larger than output vocab that were clamped to the vocab
    bar = torch.nn.functional.pad(scores, (0, 1))
    if torch.device(device).type == "cuda":
        torch.cuda.synchronize()
    dt = time.time() - t0
    times.append(dt)

print(f"Took {torch.mean(torch.tensor(times)):.2e} s on average over {N_repeat} runs")

model = AutoModelForCausalLM.from_pretrained("meta-llama/Meta-Llama-3-8B-Instruct", device_map=device)

times = []
for i in range(N_repeat):
    if torch.device(device).type == "cuda":
        torch.cuda.synchronize()
    t0 = time.time()
    with torch.no_grad():
        foo = model(torch.randint(0, vocab_size, (batch_size, 1), dtype=int, device=device))
    if torch.device(device).type == "cuda":
        torch.cuda.synchronize()
    dt = time.time() - t0
    times.append(dt)

print(f"A simple decoding forward took {torch.mean(torch.tensor(times)):.2e} s on average over {N_repeat} runs")
```

**Comment by RaphaelKreft** [comment]
Hello @Cyrilvallez, many thanks for incorperating these logit processor fixes to enable a simpler Apertus 1.5 implementation. While this fixes beamsearch and repetition-penalty, there are some other cases I found after some research. I attach a reproducer script for those potential failing use cases. 

Please let me know whether those are still in scope to be fixed or can be ignored. 
Because removing the padding of logits in our [Apertus PR](https://github.com/huggingface/transformers/pull/47662#issuecomment-5667280115) in the current state would mean, Apertus 1.5  will not work in the use cases covered by the reproducer. 

[reproduce_generation_padding_remaining.py](https://github.com/user-attachments/files/32299561/reproduce_generation_padding_remaining.py)
[generation_padding_remaining_report.md](https://github.com/user-attachments/files/32299562/generation_padding_remaining_report.md)


PS: I will also copy this message to our PR for better visibility.

**Comment by Cyrilvallez** [comment]
Hey @RaphaelKreft! All of those cases will never arise in practice and kind of only theoretical issues found by AI - no one will actually use them in practice. `NoRepeatNGramLogitsProcessor` was already very very niche, so those here are not worth fixing.
We can (and will) of course revisit later on if someone genuinely wants/needs to use those with Apertus, but for now not worth spending time at all!
